### ISSUE TITLE
The understanding of the formula in the paper and the details of the training in the code

### ISSUE THREAD

**Issue by yanchenyuxia** [hendrycks/outlier-exposure#19]
Hello, I am a graduate student in school and I just read this paper of yours. I want to ask you some questions that I don’t understand.
First of all, I don't understand the latter part of the formula in the paper.I don’t know what it means.Is f(x') the predicted output of the model on OOD data?Is the cross entropy calculated for the second half?
Second,
loss += 0.5 * -(x[len(in_set[0]):].mean(1) - torch.logsumexp(x[len(in_set[0]):], dim=1)).mean()
I don't understand how this formula shows that the obtained is cross entropy?And where is softmax used in the first half（x[len(in_set[0]):].mean(1)） of the formula?How does the second half of the formula（torch.logsumexp(x[len(in_set[0]):], dim=1)）  represent uniform distribution?
Thansk,sir!

**Comment by xfffrank** [comment]
Hi, let me try answering your questions. @yanchenyuxia 

1. f(x') is the predictions on OOD data. Since the OOD data doesn't belong to the training classes, in the second half, it's calculating the cross entropy from f(x') to the uniform distribution.

2. For your second question, you may refer to the answer in #18 .

**Comment by yanchenyuxia** [comment]
> Hi, let me try answering your questions. @yanchenyuxia
> 
> 1. f(x') is the predictions on OOD data. Since the OOD data doesn't belong to the training classes, in the second half, it's calculating the cross entropy from f(x') to the uniform distribution.
> 2. For your second question, you may refer to the answer in #18 .

I sent you an email, your Google mailbox! The phone number on my mailbox name is my WeChat ID. I still don't understand these questions, so please guide me. Thank you, sir!

**Comment by captainfffsama** [comment]
It seems that answer is not clearly,I try show my understand about why `loss += 0.5 * -(x[len(in_set[0]):].mean(1) - torch.logsumexp(x[len(in_set[0]):], dim=1)).mean()` is compute cross-entropy from softmax distribution to uniform distribution.

we know cross-entropy represent is:  

![f1](https://user-images.githubusercontent.com/40843698/120983101-f6ddd400-c7ab-11eb-83a8-f5304c705505.png)

now p(x) is unigorm distribution an q(x) is softmax distribution.

so we rewrite Formula 1 like this:
![f2](https://user-images.githubusercontent.com/40843698/120983220-15dc6600-c7ac-11eb-9ab3-b1080390b060.png)

and `x[len(in_set[0]):].mean(1)` is  first term and `torch.logsumexp(x[len(in_set[0]):], dim=1)` is second term
